In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv("substancias_remineralizadoras.csv")

df.head()

,Autor,Ano,Substancia 1,Substancia 2,Substancia 3,Substância 4,Substancia 5,Concentração substancia 1,Concentração substancia 2,Concentração substancia 3,...,Area_de_superficie_polar,Teste_microdureza,Tempo Teste EDX (dias),Teste_EDX,Tempo microscopia (dias),Microscopia_eletrônica (micrometros) µm²,Microscopia_eletrônica (classificação),Mineraliza ou não,Doi,Justificativa
0,"Goldberg, Michel et al.",2011,colageno_tipo_1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,https://pmc.ncbi.nlm.nih.gov/articles/PMC3360947/,NaN
1,"Goldberg, Michel et al.",2011,colageno_tipo_3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN
2,"Goldberg, Michel et al.",2011,hidroxiapatita,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,556.5,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN
3,"Goldberg, Michel et al.",2011,glicoproteina_Da_dentina,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN
4,"Goldberg, Michel et al.",2011,sialoproteina_ossea,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN


In [3]:
df.shape

(109, 57)

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 109 entries, 0 to 108
Data columns (total 57 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Autor                                     109 non-null    str    
 1   Ano                                       109 non-null    int64  
 2   Substancia 1                              109 non-null    str    
 3   Substancia 2                              35 non-null     str    
 4   Substancia 3                              8 non-null      str    
 5   Substância 4                              5 non-null      str    
 6   Substancia 5                              1 non-null      str    
 7   Concentração substancia 1                 24 non-null     str    
 8   Concentração substancia 2                 15 non-null     str    
 9   Concentração substancia 3                 8 non-null      str    
 10  Concentração substancia 4                 5 non-n

In [5]:
df.columns

Index(['Autor', 'Ano', 'Substancia 1', 'Substancia 2', 'Substancia 3',
       'Substância 4', 'Substancia 5', 'Concentração substancia 1',
       'Concentração substancia 2', 'Concentração substancia 3',
       'Concentração substancia 4', 'Concentração substancia 5',
       'Concentração g substancia 1', 'Porcentagem substancia 1',
       'Porcentagem substancia 2', 'Frequencia de aplicação/dia',
       'Tempo de aplicação (min)', 'Concentração Composição 1', 'Composição 1',
       'Concentração Composição 2', 'Composição 2',
       'Concentração Composição 3', 'Composição 3', 'Concentração comp 4',
       'Composição 4', 'Concentração comp 5', 'Composição 5',
       'Concentração comp 6', 'Composição 6', 'Composição 7', 'Composição 8',
       'Composição 9', 'Composição 10', 'Composição 11', 'Composição 12',
       'Composição 13', 'Composição 14', 'Composição 15', 'Composição 16',
       'Sequência de aminoácidos', 'Sequencia de aminoacidos 2',
       'Numero de aminoacidos', 'Formu

In [6]:
#conferir distribuição da variável alvo
df["Mineraliza ou não"].value_counts()

Mineraliza ou não
1    72
0    37
Name: count, dtype: int64

In [7]:
df["Mineraliza ou não"].value_counts(normalize=True) * 100

Mineraliza ou não
1    66.055046
0    33.944954
Name: proportion, dtype: float64

In [8]:
# Quantidade de valores ausentes por coluna
valores_ausentes = df.isnull().sum().sort_values(ascending=False)

valores_ausentes

Justificativa                               109
Concentração substancia 5                   108
Sequencia de aminoacidos 2                  108
Composição 12                               108
Composição 13                               108
Composição 14                               108
Composição 16                               108
Composição 15                               108
Substancia 5                                108
Concentração comp 6                         107
Porcentagem substancia 2                    105
Composição 9                                105
Microscopia_eletrônica (micrometros) µm²    105
Concentração comp 5                         105
Composição 11                               105
Composição 10                               105
Substância 4                                104
Concentração substancia 4                   104
Frequencia de aplicação/dia                 104
pH                                          103
Area_de_superficie_polar                

In [9]:
# Tabela com quantidade e percentual de valores ausentes
missing = pd.DataFrame({
    "valores_ausentes": df.isnull().sum(),
    "percentual_ausente": (df.isnull().sum() / len(df)) * 100
})

missing = missing.sort_values(by="percentual_ausente", ascending=False)

missing

,valores_ausentes,percentual_ausente
Justificativa,109,100.000000
Concentração substancia 5,108,99.082569
Sequencia de aminoacidos 2,108,99.082569
Composição 12,108,99.082569
Composição 13,108,99.082569
Composição 14,108,99.082569
Composição 16,108,99.082569
Composição 15,108,99.082569
Substancia 5,108,99.082569
Concentração comp 6,107,98.165138


## Análise inicial da variável alvo

A variável alvo utilizada foi "Mineraliza ou não", na qual:

- 1 representa substâncias com potencial/evidência de mineralização ou remineralização.
- 0 representa substâncias sem potencial/evidência suficiente.

A base possui 109 registros, sendo 72 classificados como positivos e 37 como negativos. Portanto, há um desbalanceamento moderado entre as classes, com predominância da classe positiva.

## Análise de valores ausentes

A análise de valores ausentes mostrou que muitas variáveis apresentam alto percentual de dados faltantes. Isso indica que a base é esparsa, possivelmente devido à heterogeneidade dos artigos utilizados na construção do dataset.

A variável alvo "Mineraliza ou não" está completa, com 109 registros preenchidos. No entanto, várias variáveis físico-químicas e experimentais possuem mais de 70% de valores ausentes, o que limita sua utilização direta no modelo.

Por esse motivo, a seleção de variáveis para o primeiro modelo será feita de forma conservadora, priorizando colunas com maior preenchimento e evitando variáveis que possam causar vazamento de dados.

In [10]:
alvo = "Mineraliza ou não"

colunas_identificacao = [
    "Autor",
    "Ano",
    "Doi",
    "Justificativa"
]

colunas_experimentais = [
    "Teste_microdureza",
    "Tempo Teste EDX (dias)",
    "Teste_EDX",
    "Tempo microscopia (dias)",
    "Microscopia_eletrônica (micrometros) µm²",
    "Microscopia_eletrônica (classificação)"
]

In [11]:
colunas_candidatas = [
    col for col in df.columns
    if col not in colunas_identificacao
    and col not in colunas_experimentais
    and col != alvo
]

colunas_candidatas

['Substancia 1',
 'Substancia 2',
 'Substancia 3',
 'Substância 4',
 'Substancia 5',
 'Concentração substancia 1',
 'Concentração substancia 2',
 'Concentração substancia 3',
 'Concentração substancia 4',
 'Concentração substancia 5',
 'Concentração g substancia 1',
 'Porcentagem substancia 1',
 'Porcentagem substancia 2',
 'Frequencia de aplicação/dia',
 'Tempo de aplicação (min)',
 'Concentração Composição 1',
 'Composição 1',
 'Concentração Composição 2',
 'Composição 2',
 'Concentração Composição 3',
 'Composição 3',
 'Concentração comp 4',
 'Composição 4',
 'Concentração comp 5',
 'Composição 5',
 'Concentração comp 6',
 'Composição 6',
 'Composição 7',
 'Composição 8',
 'Composição 9',
 'Composição 10',
 'Composição 11',
 'Composição 12',
 'Composição 13',
 'Composição 14',
 'Composição 15',
 'Composição 16',
 'Sequência de aminoácidos',
 'Sequencia de aminoacidos 2',
 'Numero de aminoacidos',
 'Formula_molecular',
 'pH',
 'Peso_molecular (g/mol)',
 'Aceptores_de_ligacoes_de_hidr

In [12]:
missing_candidatas = missing.loc[colunas_candidatas].sort_values(
    by="percentual_ausente",
    ascending=False
)

missing_candidatas

,valores_ausentes,percentual_ausente
Substancia 5,108,99.082569
Concentração substancia 5,108,99.082569
Composição 16,108,99.082569
Sequencia de aminoacidos 2,108,99.082569
Composição 15,108,99.082569
Composição 14,108,99.082569
Composição 13,108,99.082569
Composição 12,108,99.082569
Concentração comp 6,107,98.165138
Concentração comp 5,105,96.330275


In [13]:
colunas_com_menos_ausentes = missing_candidatas[
    missing_candidatas["percentual_ausente"] <= 85
].index.tolist()

colunas_com_menos_ausentes

['Sequência de aminoácidos',
 'Concentração Composição 1',
 'Composição 5',
 'Concentração substancia 1',
 'Composição 4',
 'Substancia 2',
 'Composição 3',
 'Composição 2',
 'Composição 1',
 'Substancia 1']

In [14]:
df_modelo = df[colunas_com_menos_ausentes + [alvo]].copy()

df_modelo.head()

,Sequência de aminoácidos,Concentração Composição 1,Composição 5,Concentração substancia 1,Composição 4,Substancia 2,Composição 3,Composição 2,Composição 1,Substancia 1,Mineraliza ou não
0,MFSFVDLRLLLLLGATALLTHGQEDIPEVSCIHNGLRVPNGETWKP...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,colageno_tipo_1,0
1,MMSFVQKGSWLLLALLHPTIILAQQEAVEGGCSHLGQSYADRDVWK...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,colageno_tipo_3,0
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,hidroxiapatita,0
3,IPVPKGNFPKGMAREANTGGDDPKTTMSPSTSSVSSGNGSVTNSGI...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,glicoproteina_Da_dentina,0
4,MKTALILLSILGMACAFSMKNLHRRVKIEDSEENGVFKYRPRYYLY...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,sialoproteina_ossea,0


In [15]:
df_modelo.shape

(109, 11)

In [16]:
colunas_com_menos_ausentes

['Sequência de aminoácidos',
 'Concentração Composição 1',
 'Composição 5',
 'Concentração substancia 1',
 'Composição 4',
 'Substancia 2',
 'Composição 3',
 'Composição 2',
 'Composição 1',
 'Substancia 1']

In [17]:
alvo = "Mineraliza ou não"

features_modelo_1 = [
    'Sequência de aminoácidos',
    'Concentração Composição 1',
    'Composição 5',
    'Concentração substancia 1',
    'Composição 4',
    'Substancia 2',
    'Composição 3',
    'Composição 2',
    'Composição 1',
    'Substancia 1'
]

df_modelo = df[features_modelo_1 + [alvo]].copy()

df_modelo.head()

,Sequência de aminoácidos,Concentração Composição 1,Composição 5,Concentração substancia 1,Composição 4,Substancia 2,Composição 3,Composição 2,Composição 1,Substancia 1,Mineraliza ou não
0,MFSFVDLRLLLLLGATALLTHGQEDIPEVSCIHNGLRVPNGETWKP...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,colageno_tipo_1,0
1,MMSFVQKGSWLLLALLHPTIILAQQEAVEGGCSHLGQSYADRDVWK...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,colageno_tipo_3,0
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,hidroxiapatita,0
3,IPVPKGNFPKGMAREANTGGDDPKTTMSPSTSSVSSGNGSVTNSGI...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,glicoproteina_Da_dentina,0
4,MKTALILLSILGMACAFSMKNLHRRVKIEDSEENGVFKYRPRYYLY...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,sialoproteina_ossea,0


In [18]:
df_modelo.shape

(109, 11)

In [19]:
df_modelo.isnull().sum().sort_values(ascending=False)

Sequência de aminoácidos     92
Concentração Composição 1    90
Composição 5                 88
Concentração substancia 1    85
Composição 4                 82
Substancia 2                 74
Composição 3                 74
Composição 2                 61
Composição 1                 60
Substancia 1                  0
Mineraliza ou não             0
dtype: int64

In [20]:
df_modelo[features_modelo_1] = df_modelo[features_modelo_1].fillna("Não informado")

In [21]:
df_modelo.isnull().sum()

Sequência de aminoácidos     0
Concentração Composição 1    0
Composição 5                 0
Concentração substancia 1    0
Composição 4                 0
Substancia 2                 0
Composição 3                 0
Composição 2                 0
Composição 1                 0
Substancia 1                 0
Mineraliza ou não            0
dtype: int64

In [22]:
df_modelo[alvo].value_counts()

Mineraliza ou não
1    72
0    37
Name: count, dtype: int64

## Seleção inicial de variáveis

Para o primeiro modelo exploratório, foram selecionadas as colunas com percentual de valores ausentes menor ou igual a 85%, excluindo variáveis de identificação, referência bibliográfica e variáveis experimentais que poderiam causar vazamento de dados.

Como a maioria das variáveis selecionadas é categórica ou textual, os valores ausentes foram preenchidos com a categoria "Não informado". Essa estratégia foi adotada para evitar a exclusão de amostras, considerando o tamanho reduzido da base.